src/main/java/graphql/execution/instrumentation/ChainedInstrumentation.java

```diff
@@ -67,21 +66,48 @@ public List<Instrumentation> getInstrumentations() {
         return instrumentations;
     }
 
-    protected InstrumentationState getSpecificState(Instrumentation instrumentation, InstrumentationState parametersInstrumentationState) {
-        ChainedInstrumentationState chainedInstrumentationState = (ChainedInstrumentationState) parametersInstrumentationState;
-        return chainedInstrumentationState.getState(instrumentation);
-    }
-
-    private <T> InstrumentationContext<T> chainedCtx(Function<Instrumentation, InstrumentationContext<T>> mapper) {
+    private <T> InstrumentationContext<T> chainedCtx(InstrumentationState state, BiFunction<Instrumentation, InstrumentationState, InstrumentationContext<T>> mapper) {
         // if we have zero or 1 instrumentations (and 1 is the most common), then we can avoid an object allocation
         // of the ChainedInstrumentationContext since it won't be needed
         if (instrumentations.isEmpty()) {
             return SimpleInstrumentationContext.noOp();
         }
+        ChainedInstrumentationState chainedInstrumentationState = (ChainedInstrumentationState) state;
         if (instrumentations.size() == 1) {
-            return mapper.apply(instrumentations.get(0));
+            return mapper.apply(instrumentations.get(0), chainedInstrumentationState.getState(0));
+        }
+        return new ChainedInstrumentationContext<>(chainedMapAndDropNulls(chainedInstrumentationState, mapper));
+    }
+
+    private <T> T chainedInstrument(InstrumentationState state, T input, ChainedInstrumentationFunction<Instrumentation, InstrumentationState, T, T> mapper) {
+        ChainedInstrumentationState chainedInstrumentationState = (ChainedInstrumentationState) state;
+        for (int i = 0; i < instrumentations.size(); i++) {
+            Instrumentation instrumentation = instrumentations.get(i);
+            InstrumentationState specificState = chainedInstrumentationState.getState(i);
+            input = mapper.apply(instrumentation, specificState, input);
+        }
+        return input;
+    }
+
+    protected <T> ImmutableList<T> chainedMapAndDropNulls(InstrumentationState state, BiFunction<Instrumentation, InstrumentationState, T> mapper) {
+        ImmutableList.Builder<T> result = ImmutableList.builderWithExpectedSize(instrumentations.size());
+        for (int i = 0; i < instrumentations.size(); i++) {
+            Instrumentation instrumentation = instrumentations.get(i);
+            InstrumentationState specificState = ((ChainedInstrumentationState) state).getState(i);
+            T value = mapper.apply(instrumentation, specificState);
+            if (value != null) {
+                result.add(value);
+            }
+        }
+        return result.build();
```

Comment:
I think this is going to be a problem with the drop nulls in list form

An instrumentation is allow to provide a null `state` so if we drop them in the list like here I suspect for a list of N instrumentations we will have < N states and hence the states will not line up

We should create a test that shows that a null state is handled

Reply:
It's covered by the existing tests, but I added an explicit case to the existing tests:

<img width="1186" alt="Screenshot 2024-02-20 at 5 24 30 pm" src="https://github.com/graphql-java/graphql-java/assets/1479220/eea68ebd-a16a-4f35-a10c-9fcfa7063c85">

The only place where it's used where nulls weren't already being filtered is `beginDeferredField` and it looks like that's desirable anyway.

Reply:
OK I have had a look at the existing code now and I can see how null instrumentation states wont matter - nor will null InstrumentationContexts

Thanks though for the extra explanation

Reply:
OK I have had a look at the existing code now and I can see how null instrumentation states wont matter - nor will null InstrumentationContexts

Thanks though for the extra explanation